libs/runtime/ebpf_object.c

```diff
@@ -260,10 +183,35 @@ ebpf_object_initialize(
     object->type = object_type;
     object->free_function = free_function;
     object->get_program_type = get_program_type_function;
+    object->id = ebpf_interlocked_increment_int32((volatile int32_t*)&_ebpf_next_id);
+    // Skip invalid IDs.
+    while (object->id == 0 || object->id == EBPF_ID_NONE) {
+        object->id = ebpf_interlocked_increment_int32((volatile int32_t*)&_ebpf_next_id);
+    }
     ebpf_list_initialize(&object->object_list_entry);
     _update_reference_history(object, EBPF_OBJECT_CREATE, file_id, line);
 
-    return _ebpf_object_tracking_list_insert(object, file_id, line);
+    ebpf_result_t result;
+    ebpf_id_entry_t entry = {.reference_count = 1, .type = object_type, .object = object};
+
+    // Use EBPF_HASH_TABLE_OPERATION_INSERT so that it fails if the key already exists.
+    result = ebpf_hash_table_update(
+        _ebpf_id_table, (const uint8_t*)&object->id, (const uint8_t*)&entry, EBPF_HASH_TABLE_OPERATION_INSERT);
+    if (result != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+#if !defined(NDEBUG)
+    ebpf_id_entry_t* new_entry = NULL;
+    result = ebpf_hash_table_find(_ebpf_id_table, (const uint8_t*)&object->id, (uint8_t**)&new_entry);
+    if (result != EBPF_SUCCESS) {
+        __fastfail(FAST_FAIL_INVALID_REFERENCE_COUNT);
+    }
+    _update_reference_history(new_entry, EBPF_OBJECT_CREATE, file_id, line);
+#endif
+
+Done:
+    return result;
```

Comment:
Will we leak object ids if `ebpf_hash_table_update` fails, or do orphan ids not matter?